File: data-access/templates.go
package dataAccess

import (
	"github.com/webdevcody/go-mailing-list/db"
)

type Template struct {
	Id      int64
	Mjml    string
	Html    string
	Text    string
	Subject string
}

func CreateTemplate(mjml string, html string, text string, subject string) (Template, error) {
	db := db.GetDB()
	stmt, err := db.Prepare("INSERT INTO templates (mjml, html, text, subject) VALUES (?, ?, ?, ?) RETURNING id, mjml, html, text, subject")
	if err != nil {
		panic(err)
	}
	defer stmt.Close()

	var createdTemplate Template
	err = stmt.QueryRow(mjml, html, text, subject).Scan(&createdTemplate.Id,
		&createdTemplate.Mjml,
		&createdTemplate.Html,
		&createdTemplate.Text,
		&createdTemplate.Subject)
	if err != nil {
		return Template{}, err
	}

	return createdTemplate, nil
}

func UpdateTemplate(id int64, mjml string, html string, text string, subject string) (Template, error) {
	db := db.GetDB()
	stmt, err := db.Prepare("UPDATE templates SET mjml = ?, html = ?, text = ?, subject = ? WHERE id = ? RETURNING id, mjml, html, text, subject")
	if err != nil {
		panic(err)
	}
	defer stmt.Close()

	var createdTemplate Template
	err = stmt.QueryRow(mjml, html, text, subject, id).Scan(
		&createdTemplate.Id,
		&createdTemplate.Mjml,
		&createdTemplate.Html,
		&createdTemplate.Text,
		&createdTemplate.Subject)
	if err != nil {
		return Template{}, err
	}

	return createdTemplate, nil
}

func DeleteTemplate(templateId int64) error {
	db := db.GetDB()
	stmt, err := db.Prepare("DELETE FROM templates WHERE id = ?")
	if err != nil {
		return err
	}
	_, err = stmt.Exec(templateId)
	if err != nil {
		return err
	}
	return nil
}

func GetTemplate(templateId int64) (Template, error) {
	db := db.GetDB()
	stmt, err := db.Prepare("SELECT * FROM templates WHERE id = ?")
	if err != nil {
		return Template{}, err
	}
	defer stmt.Close()

	var template Template
	err = stmt.QueryRow(templateId).Scan(&template.Id, &template.Mjml, &template.Html, &template.Text, &template.Subject)
	if err != nil {
		return Template{}, err
	}

	return template, nil
}

func GetTemplates() ([]Template, error) {
	db := db.GetDB()
	results, err := db.Query("SELECT * FROM templates")
	if err != nil {
		return nil, err
	}

	var templates []Template
	for results.Next() {
		var template Template
		if err := results.Scan(&template.Id, &template.Mjml, &template.Html, &template.Text, &template.Subject); err != nil {
			return nil, err
		}
		templates = append(templates, template)
	}
	return templates, nil
}

File: src/data-access/templates.ts
import { eq } from 'drizzle-orm'
import { db } from '~/db'
import { templates, type Template } from '~/db/schema'

export const defaultMjml = `<mjml>
  <mj-body>
    <mj-section>
      <mj-column>

        <mj-image width="100px" src="/assets/img/logo-small.png"></mj-image>

        <mj-divider border-color="#F45E43"></mj-divider>

        <mj-text font-size="20px" color="#F45E43" font-family="helvetica">Hello World</mj-text>

      </mj-column>
    </mj-section>
  </mj-body>
</mjml>`

export async function listTemplates() {
  return db.select().from(templates).orderBy(templates.id)
}

export async function getTemplate(id: number) {
  return db.select().from(templates).where(eq(templates.id, id)).get()
}

export async function createTemplate() {
  return db
    .insert(templates)
    .values({
      mjml: defaultMjml,
      html: 'HTML',
      text: 'TEXT',
      subject: 'This is your email subject',
    })
    .returning()
    .get()
}

export async function updateTemplate(input: Required<Pick<Template, 'id' | 'mjml' | 'html' | 'text' | 'subject'>>) {
  return db
    .update(templates)
    .set({
      mjml: input.mjml,
      html: input.html,
      text: input.text,
      subject: input.subject,
    })
    .where(eq(templates.id, input.id))
    .returning()
    .get()
}

export async function deleteTemplate(id: number) {
  db.delete(templates).where(eq(templates.id, id)).run()
}
